### PR TITLE
DX: Provider generics phpdoc

### PR DESCRIPTION
**Pull request by GregoireHebert**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Tickets       | ~
| License       | MIT
| Doc PR        | ~

Add some phpdoc generics.

### CODE REVIEW

File: src/Doctrine/Orm/Extension/QueryResultCollectionExtensionInterface.php [hunk 1]
```diff
@@ -22,13 +22,15 @@
  *
  * @author Samuel ROZE <samuel.roze@gmail.com>
  * @author Kévin Dunglas <dunglas@gmail.com>
+ *
+ * @template T
```
Comment (dunglas) [inline src/Doctrine/Orm/Extension/QueryResultCollectionExtensionInterface.php:26]:
```suggestion
 * @template T of object
```

File: src/Doctrine/Orm/Extension/QueryResultItemExtensionInterface.php [hunk 2]
```diff
@@ -21,13 +21,15 @@
  * for specific cases such as Query alteration.
  *
  * @author Antoine BLUCHET <soyuka@gmail.com>
+ *
+ * @template T
```
Comment (dunglas) [inline src/Doctrine/Orm/Extension/QueryResultItemExtensionInterface.php:25]:
```suggestion
 * @template T of object
```

File: src/State/Pagination/PaginatorInterface.php [hunk 3]
```diff
@@ -18,6 +18,10 @@
  * current page, as an integer.
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
+ *
+ * @template T
```
Comment (dunglas) [inline src/State/Pagination/PaginatorInterface.php:22]:
```suggestion
 * @template T of object
```

File: src/State/Pagination/PartialPaginatorInterface.php [hunk 4]
```diff
@@ -17,6 +17,10 @@
  * Partial Paginator Interface.
  *
  * @author Baptiste Meyer <baptiste.meyer@gmail.com>
+ *
+ * @template T
```
Comment (dunglas) [inline src/State/Pagination/PartialPaginatorInterface.php:21]:
```suggestion
 * @template T of object
```

File: src/State/ProviderInterface.php [hunk 5]
```diff
@@ -14,19 +14,22 @@
...
 use ApiPlatform\Metadata\Operation;
+use ApiPlatform\State\Pagination\PaginatorInterface;
 
 /**
  * Retrieves data from a persistence layer.
  *
  * @author Antoine Bluchet <soyuka@gmail.com>
  * @experimental
+ *
+ * @template T
```
Comment (dunglas) [inline src/State/ProviderInterface.php:25]:
```suggestion
 * @template T of object
```

**Comment by GregoireHebert** [comment]
Review accepted @dunglas :)

File: src/State/ProviderInterface.php [hunk 6]
```diff
@@ -14,19 +14,22 @@
...
+ *
+ * @template T of object
  */
 interface ProviderInterface
 {
     /**
      * Provides data.
      *
-     * @return object|array|null
+     * @return T|PaginatorInterface<T>|iterable<T>|null
```
Comment (alanpoulain) [inline src/State/ProviderInterface.php:32]: `PartialPaginatorInterface` also?

Comment (dunglas) [reply src/State/ProviderInterface.php:32]: Maybe just this one actually

Comment (GregoireHebert) [reply src/State/ProviderInterface.php]: ✅ done.

**Review by alanpoulain — APPROVED** [review]

**Comment by alanpoulain** [comment]
Thanks Grégoire.

**Pull request merged**